### ISSUE TITLE
BulkDelete not working if the keys not exist

### ISSUE THREAD

**Issue by rengert** [dfahlander/Dexie.js#1133]
Hi there,

we try to migrate to 3 but we faced an issue with the bulkDelete function.
We want to call it with keys without checking if the items exist in the database or not. This behaviour was working in 2.x

Now we receive an error as soon as in the list of keys is a key not existing in the db.

I checked the call stack and i guess this could be the issue ( i guess the generated code):

var existingValue = existingValues[i];
var ctx = { onerror: null, onsuccess: null };
if (req.type === 'delete') {
 deleting.fire.call(ctx, key, existingValue, dxTrans);
}

In our cace "existingValue" is undefined.

Reviewing the code i faced another issue. The bulkdelete method uses the type "delete" but there is a type "deleteRange" is this correct?

Thanks in advance.

**Comment by dfahlander** [comment]
Thanks! I will look into it. Regarding deleteRange can be used if you want to delete all keys in a certain range. Dexie will use it if you write a ranged delete query based on the primary key:

```js
db.friends.where('id').between(100,150).delete();
```

But if you are listening to the "deleting" hook Dexie has to first query the existing values in order to call the hooks properly before doing the deletion. In that case there's no particular win to use the deleteRange method.

However, bulkDelete() should only fail if you pass invalid keys to it, such as undefined, null, false or objects. If you pass valid keys that just don't exist in the database, it must not fail - so I regard this issue as a bug that has to be fixed soon.

**Comment by rengert** [comment]
Thanks for you quick response. Fully understood when deleteRange is used.

I tried with hardcoded
bulkDelete([1,2,3]) => error
bulkDelete([]) => no error

At the end I see an error message that someone tries to call "id" on undefined. But this could be an issue beside dexie code?

**Comment by dfahlander** [comment]
Yes, I suppose this is a bug in Dexie that needs to be fixed. Probably only occurs when "deleting" hook is subscribed to.

**Comment by dfahlander** [comment]
I tried to reproduce this error in the latest version of Dexie but it seems to work as expected. What version of Dexie were you using?

**Comment by dfahlander** [comment]
Ok, I thought that Dexie errored in this case but did you mean that the deleting hook was called also when it was about to delete a key that was not existing?

**Comment by dfahlander** [comment]
I see! I reproduced it in a [jsitor script](https://jsitor.com/L0jEANpqC). Seems that the behavior differs between version 2.x and 3.x, as you say.

**Comment by dfahlander** [comment]
The problem is not that easily solved. We do bulkDelete a bit more optimized in version 3 where we request the existing items using IDBStore.get() instead of IDBStoreOpenCursor. This means we cannot distinguish between a stored value of undefined and a non-existing value in the database. In most situations, undefined should mean that the value did not exist but it would be incorrect for Dexie to assume that. I'm thinking of if we should list this changed behavior as a difference between Dexie 2.x and 3.x instead and leave it as it is. It would of course be possible to fix, but to the cost of a little more complex code.

How does this affect your application or library? Could you add a check for undefined in your hook or do you ever store undefined as value?

**Comment by rengert** [comment]
Thanks for your response.  I guess this is something we can go with, we do not store undefined values directly. but the complexity to update our hooks should be low enough the handle this by our own.

We are using dexie for a complex offline sync service with many hooks to notify other services / component. 

Does the "problem" only exists for the bulkDelete and the "deleting" hook? or do we need to check code for bulkGet and read hooks as well?

**Comment by dfahlander** [comment]
bulkGet is new in 3.x. Reading hook has the same behavior as it had in 2.x but that involves passing unrefined through it if a get or bulkGet results in undefined.

**Comment by rengert** [comment]
Great, we fixed it as suggested by updating our hooks.
I think it is a great idea to update documentation ... i see you already added the flag.
Thanks!!
